Rails N + 2 queries (#106)
rootstrap/blog: content/Rails N + 2 queries - Bruno Vezoli.md

## Rails N + 2 queries

We usually try to solve performance problems by using `#includes` to get rid of N + 1 queries but this doesn't
always get rid of them; in fact it can even create more queries under some circumstances. This post
shows some examples of when that happens and how to deal with it.

![Forest with bridge](images/forest_bridge.jpg)

### Your typical N + 1 query problem

We learn how to deal with ActiveRecord performance issues with very plain examples, most blog posts show models that
look something like this:

```ruby
class Post < ApplicationRecord
  has_many :comments
end

class Comment < ApplicationRecord
  belongs_to :post
end

> ```

Author: Hell yeah, I love Carbon. I use it on presentations, though I don't think it's possible to use it here right?

Followed by your standard `irb` example:

```ruby
irb> posts = Post.all
irb> all_comments = posts.map { |p| p.comments }
```

> Which generates the following queries, where we can clearly see an N + 1 query since it needs to fetch comments for each
> post:

Reviewer: As this is the gist of N+1, I'll complete this sentence with something emphasizing that.

```suggestion
Which generates the following queries where we can clearly see the infamous N+1 problem: one query for retrieving the posts and N additional queries for retrieving the comments, where N is the number of posts.
```

Post Load (0.5ms)  SELECT  "posts".* FROM "posts"
Comment Load (0.8ms)  SELECT  "comments".* FROM "comments" WHERE "comments"."post_id" = $1  [["post_id", 1]]
Comment Load (0.2ms)  SELECT  "comments".* FROM "comments" WHERE "comments"."post_id" = $1  [["post_id", 2]]
Comment Load (0.2ms)  SELECT  "comments".* FROM "comments" WHERE "comments"."post_id" = $1  [["post_id", 3]]
Comment Load (0.2ms)  SELECT  "comments".* FROM "comments" WHERE "comments"."post_id" = $1  [["post_id", 4]]
Comment Load (0.6ms)  SELECT  "comments".* FROM "comments" WHERE "comments"."post_id" = $1  [["post_id", 5]]
```

And that the solution is to preload the data by calling the `#includes` method over the posts collection

```ruby
irb> posts = Post.includes(:comments)
irb> all_comments = posts.map { |p| p.comments }
```

Which in turn produces the following, much optimized, queries:

```
Post Load (0.5ms)  SELECT  "posts".* FROM "posts"
Comment Load (1.4ms)  SELECT "comments".* FROM "comments" WHERE "comments"."post_id" IN ($1, $2, $3, $4, $5)  [["post_id", 1], ["post_id", 2], ["post_id", 3], ["post_id", 4], ["post_id", 5]]
```

And so we start using `#includes` on all our ActiveRecord queries and move on.

> ### A little ways down the road

Author: Yeah like time goes by and you will have to change code and this issue will reappear if you are not careful. The title is also a reference to How I met your mother 😅

Comment Load (0.4ms)  SELECT  "comments".* FROM "comments" WHERE "comments"."post_id" = $1 ORDER BY "comments"."likes" DESC LIMIT $2  [["post_id", 5], ["LIMIT", 1]]
```

And here is where `has_one` comes to the rescue, let's modify our Posts model one last time

```ruby
class Post < ApplicationRecord
  has_many :comments
  has_many :uncensored_comments, -> { where(censored: false) }, class_name: 'Comment'
  # rails will automatically limit the number of records for us
  has_one :most_liked_comment, -> { order(likes: :desc) }, class_name: 'Comment'
end
```

Bullet now complains about data not being preload us and gives us the solution to our problems

```ruby
irb> posts = Post.includes(:most_liked_comment)
irb> most_liked_comments = posts.map { |p| p.most_liked_comment }
```

And we get neat SQL queries once more

```
Post Load (0.4ms)  SELECT "posts".* FROM "posts"
Comment Load (0.4ms)  SELECT "comments".* FROM "comments" WHERE "comments"."post_id" IN ($1, $2, $3, $4, $5) ORDER BY "comments"."likes" DESC  [["post_id", 1], ["post_id", 2], ["post_id", 3], ["post_id", 4], ["post_id", 5]]
```

> Notice though that in this case we don't get the limit on the query and so Rails loads all these comments on memory and then

Reviewer 2: `LIMIT`

`comments` association. This requires us to add one more line to our Posts model:

```ruby
class Post < ApplicationRecord
  has_many :comments
  # we need to specify a new name, a lambda to filter the comments and the model class name
  has_many :uncensored_comments, -> { where(censored: false) }, class_name: 'Comment'
end
```

And change our code to get the comments using the association:

```ruby
irb> posts = Post.all
irb> all_comments = posts.map { |p| p.uncensored_comments }
```

Bingo! We get the following warning from bullet:

```
GET /posts
USE eager loading detected
  Post => [:uncensored_comments]
  Add to your query: .includes([:uncensored_comments])
```

And sure enough if we add that preload

```ruby
irb> posts = Post.includes(:uncensored_comments)
irb> all_comments = posts.map { |p| p.uncensored_comments }
```

We no longer get the warning and our queries are optimized:

```
Post Load (0.5ms)  SELECT  "posts".* FROM "posts"
Comment Load (0.9ms)  SELECT "comments".* FROM "comments" WHERE "comments"."censored" = $1 AND "comments"."post_id" IN ($2, $3, $4, $5, $6)  [["censored", false], ["post_id", 1], ["post_id", 2], ["post_id", 3], ["post_id", 4], ["post_id", 5]]
```

> ### Caveats

Reviewer 3: ```suggestion
### Evaluating performance optimizations
```

As with most performance optimizations you should really measure and evaluate the changes you are about to make. It doesn't

> really make sense to add an association to your models every time you want to preload, sometimes is better to have a small

Reviewer 3: ```suggestion
really make sense to add an association to your models every time you want to preload, sometimes it is better to have a small
```
